File: frontend_web/src/features/specialist-dashboard/utils/specialistAiReportGeneration.js
export const AI_REPORT_TYPES = {
  WEEKLY: "weekly",
  MONTHLY: "monthly",
};

export const AI_REPORT_GENERATION_VALIDATION_KEYS = {
  PATIENT_REQUIRED: "specialist.reports.generate.errors.patientRequired",
  REPORT_TYPE_REQUIRED: "specialist.reports.generate.errors.reportTypeRequired",
  PERIOD_START_REQUIRED: "specialist.reports.generate.errors.periodStartRequired",
  PERIOD_END_REQUIRED: "specialist.reports.generate.errors.periodEndRequired",
  PERIOD_START_AFTER_END: "specialist.reports.generate.errors.periodStartAfterEnd",
  PERIOD_NOT_ENDED: "specialist.reports.generate.errors.periodNotEnded",
};

const VALIDATION_FALLBACKS = {
  [AI_REPORT_GENERATION_VALIDATION_KEYS.PATIENT_REQUIRED]: "Patient is required.",
  [AI_REPORT_GENERATION_VALIDATION_KEYS.REPORT_TYPE_REQUIRED]: "Report type must be weekly or monthly.",
  [AI_REPORT_GENERATION_VALIDATION_KEYS.PERIOD_START_REQUIRED]: "Start date is required.",
  [AI_REPORT_GENERATION_VALIDATION_KEYS.PERIOD_END_REQUIRED]: "End date is required.",
  [AI_REPORT_GENERATION_VALIDATION_KEYS.PERIOD_START_AFTER_END]: "period_start cannot be after period_end",
  [AI_REPORT_GENERATION_VALIDATION_KEYS.PERIOD_NOT_ENDED]: "Cannot generate report for a period that has not ended yet",
};

function translateKey(t, key, fallback) {
  if (typeof t === "function") {
    const translated = t(key);
    if (translated && translated !== key) {
      return translated;
    }
  }
  return fallback;
}

function readDateParts(value) {
  if (value instanceof Date && !Number.isNaN(value.getTime())) {
    return {
      year: value.getFullYear(),
      month: value.getMonth() + 1,
      day: value.getDate(),
    };
  }

  if (typeof value === "string") {
    const trimmed = value.trim();
    const match = /^(\d{4})-(\d{2})-(\d{2})$/.exec(trimmed);
    if (match) {
      return {
        year: Number(match[1]),
        month: Number(match[2]),
        day: Number(match[3]),
      };
    }
  }

  return null;
}

/** Calendar date only from local parts (no UTC conversion). */
export function specialistAiReportDateOnly(value) {
  const parts = readDateParts(value);
  if (!parts) {
    return null;
  }
  return new Date(parts.year, parts.month - 1, parts.day);
}

/** YYYY-MM-DD from local date parts. */
export function formatSpecialistAiReportDate(value) {
  const parts = readDateParts(value);
  if (!parts) {
    return "";
  }
  const month = String(parts.month).padStart(2, "0");
  const day = String(parts.day).padStart(2, "0");
  return `${parts.year}-${month}-${day}`;
}

export function defaultWeeklyAiReportPeriod(now = new Date()) {
  const end = specialistAiReportDateOnly(now);
  const start = new Date(end.getFullYear(), end.getMonth(), end.getDate() - 6);
  return {
    start: formatSpecialistAiReportDate(start),
    end: formatSpecialistAiReportDate(end),
  };
}

export function defaultMonthlyAiReportPeriod(now = new Date()) {
  const end = specialistAiReportDateOnly(now);
  const start = new Date(end.getFullYear(), end.getMonth(), end.getDate() - 29);
  return {
    start: formatSpecialistAiReportDate(start),
    end: formatSpecialistAiReportDate(end),
  };
}

export function defaultPeriodForSpecialistAiReportType(reportType, now = new Date()) {
  if (reportType === AI_REPORT_TYPES.MONTHLY) {
    return defaultMonthlyAiReportPeriod(now);
  }
  return defaultWeeklyAiReportPeriod(now);
}

export function resolveAiReportGeneratePath(reportType) {
  if (reportType === AI_REPORT_TYPES.MONTHLY) {
    return "/ai/reports/generate-monthly";
  }
  if (reportType === AI_REPORT_TYPES.WEEKLY) {
    return "/ai/reports/generate-weekly";
  }
  return null;
}

export function normalizeAiReportLanguage(value) {
  const normalized = typeof value === "string" ? value.trim().toLowerCase().replace(/_/g, "-") : "";
  const primary = normalized.split("-")[0];
  if (primary === "ar") {
    return "ar";
  }
  return "en";
}

export function buildAiReportGeneratePayload({
  patientId,
  reportType,
  periodStart,
  periodEnd,
  language,
}) {
  return {
    patient_id: String(patientId || "").trim(),
    period_start: formatSpecialistAiReportDate(periodStart),
    period_end: formatSpecialistAiReportDate(periodEnd),
    language: normalizeAiReportLanguage(language),
    reportType,
  };
}

export function validateSpecialistAiReportGeneration({
  patientId,
  reportType,
  periodStart,
  periodEnd,
  now = new Date(),
  t = null,
} = {}) {
  const messageFor = (key) => translateKey(t, key, VALIDATION_FALLBACKS[key]);

  if (!patientId || !String(patientId).trim()) {
    return messageFor(AI_REPORT_GENERATION_VALIDATION_KEYS.PATIENT_REQUIRED);
  }

  if (reportType !== AI_REPORT_TYPES.WEEKLY && reportType !== AI_REPORT_TYPES.MONTHLY) {
    return messageFor(AI_REPORT_GENERATION_VALIDATION_KEYS.REPORT_TYPE_REQUIRED);
  }

  if (!periodStart) {
    return messageFor(AI_REPORT_GENERATION_VALIDATION_KEYS.PERIOD_START_REQUIRED);
  }

  if (!periodEnd) {
    return messageFor(AI_REPORT_GENERATION_VALIDATION_KEYS.PERIOD_END_REQUIRED);
  }

  const start = specialistAiReportDateOnly(periodStart);
  const end = specialistAiReportDateOnly(periodEnd);
  const today = specialistAiReportDateOnly(now);

  if (!start || !end || !today) {
    return messageFor(AI_REPORT_GENERATION_VALIDATION_KEYS.PERIOD_START_REQUIRED);
  }

  if (start.getTime() > end.getTime()) {
    return messageFor(AI_REPORT_GENERATION_VALIDATION_KEYS.PERIOD_START_AFTER_END);
  }

  if (end.getTime() > today.getTime()) {
    return messageFor(AI_REPORT_GENERATION_VALIDATION_KEYS.PERIOD_NOT_ENDED);
  }

  return null;
}

File: mobile_app/lib/features/dashboard/models/specialist_ai_report_generation.dart
/// Request, date, and validation helpers for Specialist AI report generation.
library;

enum SpecialistAiReportType {
  weekly,
  monthly;

  String get apiValue => name;

  /// Explicit backend path. Never inferred from title or patient.
  String get generatePath => switch (this) {
        SpecialistAiReportType.weekly => '/ai/reports/generate-weekly',
        SpecialistAiReportType.monthly => '/ai/reports/generate-monthly',
      };

  static SpecialistAiReportType? tryParse(String? value) {
    final normalized = value?.trim().toLowerCase();
    return switch (normalized) {
      'weekly' => SpecialistAiReportType.weekly,
      'monthly' => SpecialistAiReportType.monthly,
      _ => null,
    };
  }
}

class SpecialistAiReportGenerateRequest {
  const SpecialistAiReportGenerateRequest({
    required this.patientId,
    required this.type,
    required this.periodStart,
    required this.periodEnd,
    this.language = 'en',
  });

  final String patientId;
  final SpecialistAiReportType type;
  final DateTime periodStart;
  final DateTime periodEnd;
  final String language;

  /// Backend body. Does not include specialist_id.
  Map<String, dynamic> toJson() {
    return {
      'patient_id': patientId.trim(),
      'period_start': formatSpecialistAiReportDate(periodStart),
      'period_end': formatSpecialistAiReportDate(periodEnd),
      'language': normalizeAiReportLanguage(language),
    };
  }
}

String normalizeAiReportLanguage(String? value) {
  final normalized = value?.trim().toLowerCase().replaceAll('_', '-') ?? '';
  final primary = normalized.split('-').first;
  if (primary == 'ar') {
    return 'ar';
  }
  return 'en';
}

/// Calendar date only: YYYY-MM-DD from local date parts (no UTC conversion).
String formatSpecialistAiReportDate(DateTime date) {
  final year = date.year.toString().padLeft(4, '0');
  final month = date.month.toString().padLeft(2, '0');
  final day = date.day.toString().padLeft(2, '0');
  return '$year-$month-$day';
}

DateTime specialistAiReportDateOnly(DateTime date) {
  return DateTime(date.year, date.month, date.day);
}

/// Weekly default: last 7 calendar days inclusive, ending today.
/// Example: 2026-08-13 → start 2026-08-07, end 2026-08-13.
({DateTime start, DateTime end}) defaultWeeklyAiReportPeriod([DateTime? now]) {
  final end = specialistAiReportDateOnly(now ?? DateTime.now());
  final start = end.subtract(const Duration(days: 6));
  return (start: start, end: end);
}

/// Monthly default: last 30 calendar days inclusive, ending today.
/// Example: 2026-08-13 → start 2026-07-15, end 2026-08-13.
///
/// This is only the initial default. Custom ranges remain allowed.
({DateTime start, DateTime end}) defaultMonthlyAiReportPeriod([DateTime? now]) {
  final end = specialistAiReportDateOnly(now ?? DateTime.now());
  final start = end.subtract(const Duration(days: 29));
  return (start: start, end: end);
}

/// Default From/To for the selected report type (used when the type changes).
({DateTime start, DateTime end}) defaultPeriodForSpecialistAiReportType(
  SpecialistAiReportType type, [
  DateTime? now,
]) {
  return switch (type) {
    SpecialistAiReportType.weekly => defaultWeeklyAiReportPeriod(now),
    SpecialistAiReportType.monthly => defaultMonthlyAiReportPeriod(now),
  };
}

/// Client-side validation mirroring backend generate-report rules.
/// Returns null when valid.
String? validateSpecialistAiReportGeneration({
  required String? patientId,
  required SpecialistAiReportType? type,
  required DateTime? periodStart,
  required DateTime? periodEnd,
  DateTime? now,
}) {
  if (patientId == null || patientId.trim().isEmpty) {
    return 'Patient is required.';
  }
  if (type == null) {
    return 'Report type must be weekly or monthly.';
  }
  if (periodStart == null) {
    return 'Start date is required.';
  }
  if (periodEnd == null) {
    return 'End date is required.';
  }

  final start = specialistAiReportDateOnly(periodStart);
  final end = specialistAiReportDateOnly(periodEnd);
  final today = specialistAiReportDateOnly(now ?? DateTime.now());

  if (start.isAfter(end)) {
    return 'period_start cannot be after period_end';
  }
  if (end.isAfter(today)) {
    return 'Cannot generate report for a period that has not ended yet';
  }

  return null;
}
